File: hashsetDesign.py
'''
Time Complexity
add --> O(1)
remove --> O(1)
contains --> O(1)

space complexity 
storage space of 2d matrix of size 10**6 O(10**6)
'''




class MyHashSet:

    def __init__(self):
        self.max_limit = 10**3
        self.storage = [ [False]*1001 for i in range(1000)]  #initialize 2d array
        #print(self.storage)
        

    def add(self, key: int) -> None:
        bucket = key % self.max_limit         #primary array finding bucket 
        bucket_item = key // self.max_limit   #secondary array finding bucket iteem 
        # print("Add")
        # print(bucket,end=' ')
        # print(bucket_item)
        #print(self.storage[bucket][bucket_item])
        self.storage[bucket][bucket_item] = True   
        

    def remove(self, key: int) -> None:
        bucket = key % self.max_limit
        bucket_item = key // self.max_limit
        # print("Remove")
        # print(bucket,end =' ')
        # print(bucket)
        self.storage[bucket][bucket_item] = False
        
        

    def contains(self, key: int) -> bool:
        bucket = key % self.max_limit
        bucket_item = key // self.max_limit
        # print("Contains")
        # print(bucket,end =' ')
        # print(bucket_item)
        # print(self.storage[bucket][bucket_item])
        return self.storage[bucket][bucket_item]
            
        


# Your MyHashSet object will be instantiated and called as such:
# obj = MyHashSet()
# obj.add(key)
# obj.remove(key)
# param_3 = obj.contains(key)

File: HashSet_Design.java
// Time Complexity : O(1) bcz we should ignore constructor. we only have to consider user oriented methods.
// Space Complexity : O(n)
// With double hasing technique we are designing hashset, here we are creating 2D array because when keys are 2,1000000 we are going create 1D array of 1 to 1000000 whereas in 2d array we are going to create 1000 and just in the block 0 and 1 we are going to subArray , which is so efficent
class MyHashSet {

    private boolean[][] storage;
    private int buckets;
    private int bucketItems;
    public MyHashSet() {
        buckets= 1000;
        bucketItems = 1001; // edge case for 0 and million with hash1 function is going to give the same value
        storage = new boolean[buckets][];
    }
    
    public int hash1(int key){
        return key % buckets;
    }
    
    public int hash2(int key){
        return key / bucketItems;
    }
    
    public void add(int key) {
        int bucket =  hash1(key);
        int bucketItem = hash2(key);
        
        if(storage[bucket] == null){
            storage[bucket] = new boolean[bucketItems];
        }
        storage[bucket][bucketItem] = true;
    }
    
    public void remove(int key) {
        int bucket =  hash1(key);
        int bucketItem = hash2(key);        
        if(storage[bucket] != null){
           storage[bucket][bucketItem] = false;
        }
        
    }
    
    public boolean contains(int key) {
        int bucket =  hash1(key);
        int bucketItem = hash2(key);
        
        if(storage[bucket] != null){
            return storage[bucket][bucketItem];
        }
        return false;
    }
}

/**
 * Your MyHashSet object will be instantiated and called as such:
 * MyHashSet obj = new MyHashSet();
 * obj.add(key);
 * obj.remove(key);
 * boolean param_3 = obj.contains(key);
 */